pkg/controller/cluster_pool_reconciler/cluster_pool_reconciler.go

```diff
@@ -0,0 +1,135 @@
+	}
+	c, err := controller.New(ControllerName, hiveManager, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	if err := c.Watch(
+		source.NewKindWithCache(&hivev1.ClusterPool{}, hiveManager.GetCache()),
+		clusterPoolHandler(),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for testimagestreamtagimports: %w", err)
+	}
+
+	r.log.Info("Successfully added reconciler to manager")
+	return nil
+}
+
+func clusterPoolHandler() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(o ctrlruntimeclient.Object) []reconcile.Request {
+		clusterPool, ok := o.(*hivev1.ClusterPool)
+		if !ok {
+			logrus.WithField("type", fmt.Sprintf("%T", o)).Error("Got object that was not an ClusterPool")
+			return nil
+		}
+		if clusterPool.Namespace == sourcePullSecretNamespace {
+			return nil
+		}
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{
+			Namespace: clusterPool.Namespace,
+			Name:      clusterPool.Name,
+		}}}
+	})
+}
+
+type reconciler struct {
+	log    *logrus.Entry
+	client ctrlruntimeclient.Client
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithField("request", req.String())
+	err := r.reconcile(ctx, req, log)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		log.WithError(err).Error("Reconciliation failed")
+	} else {
+		log.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+const (
+	sourcePullSecretNamespace = "ci-cluster-pool"
+	sourcePullSecretName      = "pull-secret"
+)
+
+func (r *reconciler) reconcile(ctx context.Context, req reconcile.Request, log *logrus.Entry) error {
+
+	*log = *log.WithField("namespace", req.Namespace).WithField("name", req.Namespace)
+	log.Info("Starting reconciliation")
+
+	clusterPool := &hivev1.ClusterPool{}
+	if err := r.client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: req.Namespace, Name: req.Name}, clusterPool); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Debug("The cluster pool is deleted")
+			return nil
+		}
+		return fmt.Errorf("failed to get the cluster pool %s in namespace %s: %w", req.Name, req.Namespace, err)
+	}
+
+	secret := &corev1.Secret{}
+	if err := r.client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: sourcePullSecretNamespace, Name: sourcePullSecretName}, secret); err != nil {
+		return fmt.Errorf("failed to get the secret %s in namespace %s: %w", sourcePullSecretName, sourcePullSecretNamespace, err)
+	}
+
+	target := getSecret(req.Namespace, secret)
+	if _, err := util.UpsertImmutableSecret(ctx, r.client, target); err != nil {
+		return fmt.Errorf("failed to upsert the secret %s in namespace %s: %w", target.Namespace, target.Namespace, err)
+	}
+	log.Debug("The pull secret is upserted successfully")
+	return nil
+}
+
+func getSecret(namespace string, src *corev1.Secret) *corev1.Secret {
```

Comment:
This does not "get" a secret - it clones one (why not using DeepCopy?) then replaces the namespace and adds a label.

Reply:
Updated.